addons/web/static/src/core/code_editor/code_editor.js

```diff
@@ -0,0 +1,137 @@
+/** @odoo-module */
+import { Component, onWillDestroy, onWillStart, useEffect, useRef } from "@odoo/owl";
+import { loadBundle, loadJS } from "@web/core/assets";
+import { useDebounced } from "@web/core/utils/timing";
+
+function onResized(ref, callback) {
+    const _ref = typeof ref === "string" ? useRef(ref) : ref;
+    const resizeObserver = new ResizeObserver(() => callback());
+
+    useEffect(
+        (el) => {
+            if (el) {
+                resizeObserver.observe(el);
+                return () => resizeObserver.unobserve(el);
+            }
+        },
+        () => [_ref.el]
+    );
+
+    onWillDestroy(() => {
+        resizeObserver.disconnect();
+    });
+}
+
+export class CodeEditor extends Component {
+    static template = "web.CodeEditor";
+    static components = {};
+    static props = {
+        type: {
+            type: String,
+            validate: (type) => type in CodeEditor.TYPE_LIBS,
+        },
+        value: { type: String, optional: true },
+        onChange: { type: Function, optional: true },
+        class: { type: String, optional: true },
+        theme: {
+            type: String,
+            optional: true,
+            validate: (theme) => theme === "" || theme in CodeEditor.THEMES,
+        },
+    };
+    static defaultProps = {
+        value: "",
+        onChange: () => {},
+        class: "",
+        theme: "",
+    };
+
+    static ACE_LIBS = ["/web/static/lib/ace/ace.js"];
```

Comment:
i think we should create a bundle with all ace files mentioned here, that would be simpler

Reply:
Something like that or a python bundle?
```js
onWillStart(() => {
    const aceBundle = {
        jsLibs: [
            "/web/static/lib/ace/ace.js",
            [
                "/web/static/lib/ace/mode-js.js",
                "/web/static/lib/ace/javascript_highlight_rules.js",
                "/web/static/lib/ace/mode-xml.js",
                "/web/static/lib/ace/mode-qweb.js",
                "/web/static/lib/ace/mode-scss.js",
                "/web/static/lib/ace/theme-monokai.js",
            ]
        ]
    }
    return loadBundle(aceBundle);
});
```